vllm/v1/attention/backends/cpu_attn.py

```diff
@@ -397,21 +345,33 @@ def build(self,
 
         seq_lens_cpu = common_attn_metadata.seq_lens_cpu
         seq_lens_np = seq_lens_cpu.numpy()
-        num_prompt_req = self.num_prompt_req
-        max_prefill_seq_len = seq_lens_np[:num_prompt_req].max().item(
+
+        query_start_loc_cpu = common_attn_metadata.query_start_loc_cpu
+        query_start_loc_np = query_start_loc_cpu.numpy()
+        num_scheduled_tokens = query_start_loc_np[
+            1:num_reqs + 1] - query_start_loc_np[0:num_reqs]
+        num_decode_req = 0
+        for i in range(common_attn_metadata.num_reqs):
+            if num_scheduled_tokens[i] != 1:
+                break
+            num_decode_req += 1
+        num_prompt_req = num_reqs - num_decode_req
+
+        max_prefill_seq_len = seq_lens_np[num_decode_req:num_reqs].max().item(
         ) if num_prompt_req > 0 else 0
-        max_decode_seq_len = seq_lens_np[num_prompt_req:num_reqs].max().item(
+        max_decode_seq_len = seq_lens_np[:num_decode_req].max().item(
         ) if num_prompt_req < num_reqs else 0
         self.seq_start_loc_np[0] = 0
         np.cumsum(seq_lens_np, out=self.seq_start_loc_np[1:num_reqs + 1])
 
-        query_start_loc_cpu = common_attn_metadata.query_start_loc_cpu
-        num_prefill_tokens = int(query_start_loc_cpu[num_prompt_req].item())
-        num_decode_tokens = int(query_start_loc_cpu[num_reqs].item() -
-                                num_prefill_tokens)
+        num_decode_tokens = int(query_start_loc_cpu[num_decode_req].item())
+        num_prefill_tokens = int(query_start_loc_cpu[num_reqs].item() -
+                                 num_decode_tokens)
 
         slot_mapping = common_attn_metadata.slot_mapping.long()
         block_table_tensor = common_attn_metadata.block_table_tensor
+        query_start_loc_np = query_start_loc_cpu.numpy()
+        query_start_loc_np[num_decode_req:num_reqs + 1] -= num_decode_tokens
```

Comment:
can you change these prefill/decode related things to something seperate attributes in attention metadata like `prefill_query_start_loc`, `decode_seq_lens` , etc.

Reply:
Updated some attributes. The CPU backend is under refactoring and finally will have same attributes as the FA backend.